phenotrex/ml/shap_handler.py

```diff
@@ -0,0 +1,248 @@
+
+        self._sample_names = None
+        self._used_features = None
+        self._used_shaps = None
+        self._shap_base_value = None
+        self._class_names = ['binary']
+
+    def add_feature_data(self, sample_names: np.ndarray, features: np.ndarray, shaps: np.ndarray,
+                         base_value: float = None):
+        """
+        Add a new set of feature information to the ShapHandler.
+
+        :param sample_names: an array of sample names; should globally unique.
+        :param features: a feature array of shape (n_sample_names, n_features_in_featurespace).
+                         features will be pared down to those used by the model.
+        :param shaps: a shap array of shape (n_sample_names, n_features_in_featurespace +1).
+                      shaps will be pared down to those produced by the model, mirroring the features.
+        :param base_value: If the base value has been split off the shaps before hand, pass it here.
+        :return: None
+        """
+        if base_value is None:
+            shaps, base_value = shaps[..., :-1], np.average(shaps[..., -1])
+
+        if self._shap_base_value is None:
+            self._shap_base_value = base_value
+        else:
+            assert np.allclose(self._shap_base_value, base_value), \
+                f'Incongruent base values found: {self._shap_base_value} vs. {base_value}.'
+
+        X_used = np.nan_to_num(features[:, self._used_idxs])
+        shaps_used = np.nan_to_num(shaps[..., self._used_idxs])
+
+        try:
+            X_used = X_used.toarray()
+        except AttributeError:
+            pass
+
+        if self._sample_names is None:
+            self._sample_names = sample_names
+        else:
+            self._sample_names = np.concatenate([self._sample_names, sample_names])
+
+        if self._used_features is None:
+            self._used_features = X_used
+        else:
+            self._used_features = np.concatenate([self._used_features, X_used])
+
+        if self._used_shaps is None:
+            self._used_shaps = shaps_used
+        else:
+            self._used_shaps = np.concatenate([self._used_shaps, shaps_used])
+
+    def _get_sample_index_with_name(self, sample_name: str) -> int:
+        try:
+            i = np.where(np.isin(self._sample_names, sample_name))[0][0]
+            return i
+        except (ValueError, IndexError):
+            raise ValueError('Sample label not found among saved explanations.')
+
+    def _get_feature_data(self) -> Tuple[np.ndarray, np.ndarray, np.ndarray]:
+        """
+        Concatenates and returns all currently saved features, shaps and sample names.
+        """
+        try:
+            X_agg = self._used_features.astype(float)
+            shap_agg = self._used_shaps.astype(float)
+        except (ValueError, AttributeError) as e:
+            raise RuntimeError('No explanations saved.')
+        sample_names = self._sample_names
+        return X_agg, shap_agg, sample_names
```

Comment:
I'm wondering whether we might want to return `namedtuple`s in such cases? Then one cannot mix up the order of the returned objects. Otherwise, please describe them in the docstring.